README.md
![python package](https://github.com/gbrammer/sregion/actions/workflows/python-package.yml/badge.svg)

# sregion
Parsing of IVOA S_REGION strings

The STS-C formalism is described at http://www.ivoa.net/Documents/latest/STC-S.html, though [it seems](https://github.com/astropy/regions/issues/21) that it was never adopted as an official standard.  Nevertheless, the `s_region` strings do seem to have been adopted as a sort of pseudostandard in [IVOA-compliant](https://wiki.ivoa.net/twiki/bin/view/IVOA/DCPToolsFITS) datasets / databases.

[`astropy-regions`](https://github.com/astropy/regions) would probably be a better place to put this, but I'm not interested in all of the full astropy coordinate compatibility for now.

## Examples

```python
>>> import numpy as np
>>> from sregion import SRegion

#
# Polygon string
#
>>> sr = SRegion('POLYGON 0.0 0.0 0.0 1.0 1.0 1.0 1.0 0.0')
>>> print(sr.area)
[1.0]
>>> print(sr.centroid)
[array([0.5, 0.5])]

#
# Circle string
#
>>> for i in range(4,10):
>>>     sr = SRegion('CIRCLE 10 10 1', ncircle=2**i)
>>>     print(f'ncircle={2**i:>3} {sr.area[0]/np.pi:.5f} {sr.centroid[0]}')
ncircle= 16 0.97450 [10. 10.]
ncircle= 32 0.99359 [10. 10.]
ncircle= 64 0.99839 [10. 10.]
ncircle=128 0.99960 [10. 10.]
ncircle=256 0.99990 [10. 10.]
ncircle=512 0.99997 [10. 10.]

# Circle with radius in angular units
>>> import astropy.units as u
>>> sr = SRegion('CIRCLE 10 10 1"', ncircle=256)
>>> print(f'{sr.sky_area(unit=u.arcsec**2)[0]:.5f}')
3.14128 arcsec2

#
# From WCS objects
#
>>> from astropy.wcs import WCS
>>> wcs = WCS()
>>> wcs.pixel_shape = [601,601]
>>> wcs.wcs.cdelt *= 0.1/3600
>>> wcs.wcs.crpix[1] = 300
>>> wcs.wcs.crval = [0,0]
>>> print(SRegion(wcs).sky_area())
[<Quantity 1. arcmin2>]

#
# From arrays
#
>>> x = np.array([0, 0, 1, 1])
>>> y = np.array([0, 1, 1, 0])
>>> sr = SRegion(np.array([x, y]).T)
>>> print(sr.area)
[1.0]
>>> print(sr.centroid)
[array([0.5, 0.5])]

# 
# To s_region string
#
>>> print(sr.s_region)
POLYGON 0.000000 0.000000 0.000000 1.000000 1.000000 1.000000 1.000000 0.000000

#
# To matplotlib path object(s)
#
>>> print(sr.path[0].contains_point([0.5, 0.5]))
True
>>> print(sr.path[0].contains_points([[0.5, 0.5], [2.0, 2.0]]))
[ True False]

#
# To matplotlib patch(es)
#
>>> import matplotlib.pyplot as plt
>>> fig, ax = plt.subplots(1,1,figsize=(2,2))
>>> for p in sr.patch(alpha=0.5, fc='r'):
>>>     ax.add_patch(p)
>>> ax.set_xlim(-1, 2)
>>> ax.set_ylim(*ax.get_xlim())
>>> ax.grid()

#
# To shapely polygons
# 
>>> sr.shapely
[<shapely.geometry.polygon.Polygon at 0x18055b910>]

#
# To DS9 region(s)
#
>>> for r in sr.region:
>>>    print(r)
polygon(0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000)

>>> sr.ds9_properties = 'color=red width=2'
>>> sr.label = 'my_group'
>>> for r in sr.region:
>>>    print(r)
polygon(0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000) # color=red width=2 text={my_group}

    
```


CHANGES.rst
0.1
-----
- first versioned release

1.0
-----
- includes polystr

1.1.1
-----
- New setup

sregion/tests/test_sregion.py
import numpy as np
import astropy.units as u

from .. import SRegion


def test_sregion():
    """
    Test SRegion object
    """
    # From arrays
    x = np.array([0, 0, 1, 1])
    y = np.array([0, 1, 1, 0])
    sr = SRegion(np.array([x, y]).T)

    assert(sr.N == 1)

    assert(np.allclose(sr.centroid[0], 0.5, rtol=1.e-3))

    assert(sr.area[0] == 1.0)

    # Converters
    assert(hasattr(sr.path[0], 'contains_point'))
    assert(hasattr(sr.shapely[0], 'boundary'))
    assert(hasattr(sr.patch()[0], 'get_fc'))

    regstr = ('polygon(0.000000,0.000000,0.000000,1.000000,' +
              '1.000000,1.000000,1.000000,0.000000)')
    assert(sr.region[0] == regstr)

    sr.label = 'test'
    assert(sr.region[0] == regstr + ' #  text={test}')

    # SQL string
    pstr = '((0.000,0.000),(0.000,1.000),(1.000,1.000),(1.000,0.000))'
    assert(sr.polystr(precision=3)[0] == pstr)
    
    # From s_region string
    pstr = ('POLYGON 0.000000 0.000000 0.000000 1.000000 ' +
            '1.000000 1.000000 1.000000 0.000000')

    assert(pstr == sr.s_region)

    snew = SRegion(pstr)
    assert(snew.area[0] == 1.0)

    # From polygon
    snew = SRegion(sr.shapely[0])
    assert(snew.area[0] == 1.0)
    
    # Compound regions
    x2 = np.array([0, 0, 1, 1]) + 2
    y2 = np.array([0, 1, 1, 0]) + 2
    s2 = SRegion(np.array([x2, y2]).T)

    un = sr.union(s2.shapely[0], as_polygon=True)

    assert(un.area == 2.0)

    # Multiple string
    comp = SRegion(' '.join([sr.s_region, s2.s_region]))
    assert(np.allclose(comp.area, 1.0, rtol=1.e-3))

    un = comp.union(as_polygon=True)
    assert(un.area == 2.0)

    # Intersects
    assert(sr.intersects(s2.shapely[0]) is False)

    x3 = np.array([0, 0, 1, 1]) + 0.5
    y3 = np.array([0, 1, 1, 0]) + 0.5
    s3 = SRegion(np.array([x3, y3]).T)

    assert(sr.intersects(s3.shapely[0]) is True)


def test_wrap():
    """
    """
    x = np.array([0, 0, 1, 1]) - 177
    y = np.array([0, 1, 1, 0])
    sr = SRegion(np.array([x, y]).T, wrap=True)

    assert(sr.N == 1)
    assert(np.allclose(sr.area, 1.))
    assert(np.allclose(sr.centroid, [360-177+0.5, 0.5]))


def test_circles():
    """
    Initialize from ``CIRCLE X Y R``
    """
    # CIRCLE string
    circ = SRegion('CIRCLE 5. 5. 1', ncircle=256)
    assert(np.allclose(circ.area, np.pi, rtol=1.e-3))
    assert(np.allclose(circ.centroid[0], 5., rtol=1.e-3))

    # R = 2, no units
    circ2 = SRegion('CIRCLE 5. 5. 2', ncircle=256)
    assert(np.allclose(circ2.area, 2**2*np.pi, rtol=1.e-3))
    assert(np.allclose(circ2.centroid[0], 5., rtol=1.e-3))

    # R = 1 arcsec
    cosd = np.cos(45./180*np.pi)
    csky = SRegion('CIRCLE 45. 45. 1"', ncircle=256)
    assert(np.allclose(csky.area, np.pi/3600.**2/cosd, rtol=1.e-3))
    assert(np.allclose(csky.sky_area(unit=u.arcsec**2),
                       np.pi*u.arcsec**2, rtol=1.e-3))

    # R = 1 arcmin
    cosd = np.cos(45./180*np.pi)
    csky = SRegion('CIRCLE 45. 45. 1\'', ncircle=256)
    assert(np.allclose(csky.area, np.pi/3600./cosd, rtol=1.e-3))
    assert(np.allclose(csky.sky_area(unit=u.arcsec**2),
                       3600*np.pi*u.arcsec**2, rtol=1.e-3))
    assert(np.allclose(csky.centroid[0], 45., rtol=1.e-3))

    # Sky buffer
    cosd = np.cos(45./180*np.pi)
    csky = SRegion('CIRCLE 45. 45. 0.001"', ncircle=256)
    csky.sky_buffer(1.)

    assert(np.allclose(csky.area, np.pi/cosd, rtol=1.e-3))
    assert(np.allclose(csky.sky_area(unit=u.deg**2),
                       np.pi*u.deg**2, rtol=1.e-3))


def test_whitespace():
    """
    """
    pstr = 'CIRCLE 5. 5. 1'
    for str_i in [pstr, pstr.replace(' ', '   '), '  '+pstr, pstr+'\n']:
        sr = SRegion(str_i, ncircle=256)
        assert(sr.N == 1)
        assert(np.allclose(sr.area, np.pi, rtol=1.e-3))
        assert(np.allclose(sr.centroid[0], 5., rtol=1.e-3))

    pstr = ('POLYGON 0.000000 0.000000 0.000000 1.000000 ' +
            '1.000000 1.000000 1.000000 0.000000')

    for str_i in [pstr, pstr.replace(' ', '  '), '  '+pstr, pstr+'\n']:
        sr = SRegion(str_i)
        assert(sr.N == 1)
        assert(np.allclose(sr.area, 1.))
        assert(np.allclose(sr.centroid, 0.5))


def test_fromwcs():
    """
    Initialize from `astropy.wcs.WCS`
    """
    from astropy.io.fits import Header
    import astropy.wcs as pywcs

    # From WCS
    header_string = """
CRPIX1  =                  1.0                                                  
CRPIX2  =                  1.0                                                  
CRVAL1  =                 90.0                                                  
CRVAL2  =                  0.0                                                  
CD1_1   =             -0.00001                                                  
CD1_2   =                  0.0                                                  
CD2_1   =                  0.0                                                  
CD2_2   =              0.00001                                                  
NAXIS1  =                 1001                                                  
NAXIS2  =                 1001                                                  
CTYPE1  = 'RA---TAN'                                                            
CTYPE2  = 'DEC--TAN'                                                            
RADESYS = 'ICRS    '                                                            
EQUINOX =               2000.0                                                  
LATPOLE =                    0                                                  
LONPOLE =                180.0
    """

    head = Header.fromstring(header_string.replace('\n', ''))
    wcs = pywcs.WCS(head)
    sw = SRegion(wcs)
    pixel_area = np.abs((head['NAXIS1']-1)*head['CD1_1'] *
                        (head['NAXIS2']-1)*head['CD2_2'])

    assert(np.allclose(sw.area, pixel_area, atol=1.e-5))

    assert(np.allclose(sw.sky_area(unit=u.deg**2), pixel_area*u.deg**2))
    assert(np.allclose(sw.sky_area(unit=u.arcmin**2),
                       (pixel_area*u.deg**2).to(u.arcmin**2)))


sregion/sregion.py
import numpy as np
import astropy.units as u
import astropy.wcs as pywcs


__all__ = ["SRegion"]


def _parse_circle(spli, ncircle=32):
    """
    Parse circle from split string

    Parameters
    ----------
    spli : list
        List of circle parameters ``['x', 'y', 'radius']``

    ncircle : int
        Number of points to sample around the diameter of the circle

    Returns
    -------
    poly_i : array-like
        2D xy array

    """
    if len(spli) != 3:
        msg = f'Input does not have three elements ({spli})'
        raise ValueError(msg)

    x0, y0 = np.cast[float](spli[:2])
    cosd = np.cos(y0/180*np.pi)

    r0 = spli[2]
    if r0.endswith('"'):
        scl = float(r0[:-1])/3600
    elif r0.endswith('\''):
        scl = float(r0[:-1])/60
    else:
        try:
            scl = float(r0)
        except ValueError:
            scl = 1.

        cosd = 1.

    _theta = np.linspace(0, 2*np.pi, ncircle+1)[:-1]
    _xc = np.cos(_theta)
    _yc = np.sin(_theta)

    poly_i = np.array([_xc/cosd*scl+x0, _yc*scl+y0]).T
    return poly_i


def _wrap_xy(xy):
    """
    Wrap x coordinate as an angle

    Parameters
    ----------
    xy : (M,2) array
        Coordinates

    Returns
    -------
    wxy: (M,2) array
        Array with first column wrapped at 360 degrees

    """
    from astropy.coordinates import Angle
    import astropy.units as u

    wxy = xy*1.
    ra = Angle(xy[:, 0]*u.deg).wrap_at(360*u.deg).value
    wxy[:, 0] = ra
    return wxy


def _parse_sregion(sregion, ncircle=32, **kwargs):
    """
    Parse an S_REGION string with CIRCLE or POLYGON

    Parameters
    ----------
    sregion : str
        Region string

    Returns
    -------
    poly : list
        List of 2D xy arrays

    """
    if hasattr(sregion, 'decode'):
        decoded = sregion.decode('utf-8').strip().upper()
    else:
        decoded = sregion.strip().upper()

    polyspl = decoded.replace('POLYGON', 'xxx').replace('CIRCLE', 'xxx')
    polyspl = polyspl.split('xxx')

    poly = []
    for pp in polyspl:
        if not pp:
            continue

        # strip out ()
        pp = pp.replace('(', '').replace(')', '')

        # ',' or ' ' delimiters
        if ',' in pp:
            spl = pp.strip().split(',')
        else:
            spl = pp.strip().split()

        for ip, p in enumerate(spl):
            # Find index of first float
            try:
                _pf = float(p)
                break
            except ValueError:
                continue

        if len(spl[ip:]) == 3:
            # Circle
            poly_i = _parse_circle(spl[ip:], ncircle=ncircle)
        else:
            poly_i = np.cast[float](spl[ip:]).reshape((-1, 2))

        if len(poly_i) < 2:
            continue

        poly.append(poly_i)

    return poly


class SRegion(object):
    def __init__(self, inp, label=None, wrap=True, **kwargs):
        """
        Helper class for parsing an S_REGION strings and general polygon
        tools

        Parameters
        ----------
        inp : str, (M,2) array, `~astropy.wcs.WCS`, `shapely.geometry.polygon.Polygon`

            Can be a "S_REGION" string, an array, or an `astropy.wcs.WCS`
            object from which the corners will be extracted with
            `~astropy.wcs.WCS.calc_footprint`.

        label : str
            Optional label attached to regions and patches

        wrap : bool
            Wrap first dimension as an angle between (0, 360) degrees

        """
        if isinstance(inp, str):
            self.xy = _parse_sregion(inp, **kwargs)

        elif hasattr(inp, 'sum'):
            # NDarray
            sh = inp.shape
            if inp.ndim != 2:
                raise ValueError(f'Input shape {sh} is not (M,2)')
            else:
                if inp.shape[1] != 2:
                    if inp.shape[0] != 2:
                        raise ValueError(f'Input shape {sh} is not (M,2)')
                    else:
                        inp = inp.T

            self.xy = [inp]

        elif isinstance(inp, list):
            self.xy = inp

        elif isinstance(inp, pywcs.WCS):
            self.xy = [inp.calc_footprint()]

        elif hasattr(inp, 'buffer'):
            # Shapely polygon
            if hasattr(inp, '__len__'):
                self.xy = [np.array(p.boundary.xy).T for p in inp]
            else:
                self.xy = [np.array(inp.boundary.xy).T]

        else:
            raise IOError('input must be ``str``, ``list``, or ``np.array``')

        if wrap:
            self.xy = [_wrap_xy(xy_i) for xy_i in self.xy]

        self.inp = inp
        self.ds9_properties = ''
        self.label = label

    @property
    def N(self):
        return len(self.xy)

    @property
    def centroid(self):
        return [np.mean(fp, axis=0) for fp in self.xy]

    def sky_buffer(self, buffer_deg):
        """
        Buffer polygons accounting for cos(dec)
        """
        from shapely.geometry import Polygon

        new_xy = []
        for xy, c in zip(self.xy, self.centroid):
            cosd = np.array([1, np.cos(c[1]/180*np.pi)])
            p = Polygon((xy - c)*cosd).convex_hull.buffer(buffer_deg)
            xyp = np.array(p.boundary.xy).T
            new_xy.append((xyp/cosd)+c)

        self.xy = new_xy

    @property
    def path(self):
        """
        `~matplotlib.path.Path` object
        """
        import matplotlib.path
        return [matplotlib.path.Path(fp) for fp in self.xy]

    @property
    def shapely(self):
        """
        `~shapely.geometry.Polygon` object.
        """
        from shapely.geometry import Polygon
        return [Polygon(fp).convex_hull for fp in self.xy]

    @property
    def area(self):
        """
        Area of shapely polygons
        """
        return [sh.area for sh in self.shapely]

    def sky_area(self, unit=u.arcmin**2):
        """
        Assuming coordinates provided are RA/Dec degrees, compute area
        """
        cosd = np.cos(self.centroid[0][1]/180*np.pi)
        return [(sh.area*cosd*u.deg**2).to(unit)
                for sh in self.shapely]

    def patch(self, **kwargs):
        """
        `~descartes.PolygonPatch` object
        """
        from descartes import PolygonPatch
        if 'label' not in kwargs:
            kwargs['label'] = self.label

        return [PolygonPatch(p, **kwargs) for p in self.shapely]

    def get_patch(self, **kwargs):
        """
        `~descartes.PolygonPatch` object

        ** Deprecated, use patch **
        """
        return self.patch(**kwargs)

    def union(self, shape=None, as_polygon=False):
        """
        Union of self and `shape` object.  If no `shape` provided, then
        return union of (optional) multiple components of self
        """
        if shape is None:
            un = self.shapely[0]
        else:
            un = shape

        for s in self.shapely:
            un = un.union(s)

        if as_polygon:
            return un
        else:
            return SRegion(un)

    def intersects(self, shape):
        """
        Union of self and `shape` object
        """
        test = False
        for s in self.shapely:
            test |= s.intersects(shape)

        return test

    @property
    def region(self):
        """
        Polygon string in DS9 region format
        """
        pstr = 'polygon({0})'
        if hasattr(self, 'ds9_properties'):
            tail = '{0}'.format(self.ds9_properties)
        else:
            tail = ''

        if hasattr(self, 'label'):
            if self.label is not None:
                tail += ' text={xx}'.replace('xx', self.label)

        if tail:
            tail = ' # '+tail

        return [pstr.format(','.join([f'{c:.6f}' for c in fp.flatten()]))+tail
                for fp in self.xy]

    @property
    def s_region(self):
        """
        Polygon as VO s_region
        """
        pstr = 'POLYGON {0}'
        polys = [pstr.format(' '.join([f'{c:.6f}' for c in fp.flatten()]))
                 for fp in self.xy]
        return ' '.join(polys)


    def polystr(self, precision=6):
        """
        String suitable for, e.g., SQL, e.g., 
        
        ((x1,y1),(x2,y2),...)
        """
        np.set_printoptions(precision=precision)
        
        fps = []
        for xy in self.xy:
            fmt = '({0:.pf},{1:.pf})'.replace('p', f'{precision:0}')
            fpstr = [fmt.format(*row) for row in xy]
            fps.append('('+','.join(fpstr)+')')
            
        return fps
        